mr/src/test/java/org/apache/iceberg/mr/hive/TestDeserializer.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.StandardStructObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.hive.MetastoreUtil;
+import org.apache.iceberg.types.Types;
+import org.junit.Assert;
+import org.junit.Assume;
+import org.junit.Test;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class TestDeserializer {
+  private static final Schema CUSTOMER_SCHEMA = new Schema(
+      optional(1, "customer_id", Types.LongType.get()),
+      optional(2, "first_name", Types.StringType.get())
+  );
+
+  private static final StandardStructObjectInspector CUSTOMER_OBJECT_INSPECTOR =
+      ObjectInspectorFactory.getStandardStructObjectInspector(
+          Arrays.asList("customer_id", "first_name"),
+          Arrays.asList(
+              PrimitiveObjectInspectorFactory.writableLongObjectInspector,
+              PrimitiveObjectInspectorFactory.writableStringObjectInspector
+          ));
+
+
+  @Test
+  public void testSimpleDeserialize() throws SerDeException {
+    Deserializer deserializer = new Deserializer(CUSTOMER_SCHEMA, CUSTOMER_OBJECT_INSPECTOR);
+
+    Record expected = GenericRecord.create(CUSTOMER_SCHEMA);
+    expected.set(0, 1L);
+    expected.set(1, "Bob");
+
+    Record actual = deserializer.deserialize(new Object[] { new LongWritable(1L), new Text("Bob") });
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void testDeserializeEverySupportedType() throws SerDeException {
+    // No test yet for Hive3 (Date/Timestamp creation)
+    Assume.assumeFalse(MetastoreUtil.hive3PresentOnClasspath());
+
+    Deserializer deserializer = new Deserializer(HiveIcebergTestUtils.FULL_SCHEMA,
+        HiveIcebergTestUtils.FULL_SCHEMA_OBJECT_INSPECTOR);
+
+    Record expected = HiveIcebergTestUtils.getTestRecord(false);
+    Record actual = deserializer.deserialize(HiveIcebergTestUtils.valuesForTestRecord(expected));
+
+    HiveIcebergTestUtils.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void testNullDeserialize() throws SerDeException {
+    Deserializer deserializer = new Deserializer(HiveIcebergTestUtils.FULL_SCHEMA,
+        HiveIcebergTestUtils.FULL_SCHEMA_OBJECT_INSPECTOR);
+
+    Record expected = HiveIcebergTestUtils.getNullTestRecord();
+
+    Object[] nulls = new Object[HiveIcebergTestUtils.FULL_SCHEMA.columns().size()];
+    Arrays.fill(nulls, null);
+
+    Record actual = deserializer.deserialize(nulls);
+
+    Assert.assertEquals(expected, actual);
+
+    // Check null record as well
+    Assert.assertNull(deserializer.deserialize(null));
+  }
+
+  @Test(expected = SerDeException.class)
```

Comment:
We typically prefer to use `AssertHelpers.assertThrows` rather than `expected`. When we're testing other places, we can continue making assertions after the exception is thrown that way. Here it probably doesn't matter much, but it is a good habit to be in.

Reply:
SerDeException was not a runtime exception and would have to do some extra magic to use `AssertHelpers.assertThrows`.

On the other hand I think we can convert any exception in the SerDe to SerDeException, we can throw `IllegalArgumentException`s instead and handle those outside.

Fixed